Spring fails to load token file

After OSGi features Spring fails loading the token file from the Classloader
```
[20:45:55]	[Step 2/5] Caused by: java.lang.NullPointerException: null
[20:45:55]	[Step 2/5] 	at com.vaadin.flow.server.DeploymentConfigurationFactory.getTokenFileFromClassloader(DeploymentConfigurationFactory.java:371) ~[flow-server-6.0-SNAPSHOT.jar:6.0-SNAPSHOT]
```

--- a/pom.xml
+++ b/pom.xml
@@ -11,7 +11,7 @@
     </parent>
 
     <artifactId>vaadin-spring-parent</artifactId>
-    <version>15.0-SNAPSHOT</version>
+    <version>16.0-SNAPSHOT</version>
     <packaging>pom</packaging>
 
     <name>vaadin-spring-parent</name>
@@ -44,7 +44,7 @@
         <maven.compiler.target>1.8</maven.compiler.target>
 
         <!-- These are typically overridden with BOMs -->
-        <vaadin.flow.version>5.0-SNAPSHOT</vaadin.flow.version>
+        <vaadin.flow.version>6.0-SNAPSHOT</vaadin.flow.version>
         <spring-boot.version>2.2.0.RELEASE</spring-boot.version>
 
         <!-- Additional manifest fields -->

--- a/vaadin-spring-tests/dummy-module/pom.xml
+++ b/vaadin-spring-tests/dummy-module/pom.xml
@@ -5,7 +5,7 @@
     <parent>
         <groupId>com.vaadin</groupId>
         <artifactId>vaadin-spring-tests</artifactId>
-        <version>15.0-SNAPSHOT</version>
+        <version>16.0-SNAPSHOT</version>
     </parent>
     <artifactId>dummy-module</artifactId>
     <name>Keep this module as the last one in the list</name>

--- a/vaadin-spring-tests/pom.xml
+++ b/vaadin-spring-tests/pom.xml
@@ -5,7 +5,7 @@
     <parent>
         <groupId>com.vaadin</groupId>
         <artifactId>vaadin-spring-parent</artifactId>
-        <version>15.0-SNAPSHOT</version>
+        <version>16.0-SNAPSHOT</version>
     </parent>
     <artifactId>vaadin-spring-tests</artifactId>
     <name>Vaadin Spring tests</name>

--- a/vaadin-spring-tests/test-mvc-without-endpoints/pom.xml
+++ b/vaadin-spring-tests/test-mvc-without-endpoints/pom.xml
@@ -6,7 +6,7 @@
     <parent>
         <groupId>com.vaadin</groupId>
         <artifactId>vaadin-spring-tests</artifactId>
-        <version>15.0-SNAPSHOT</version>
+        <version>16.0-SNAPSHOT</version>
     </parent>
     <artifactId>test-ts-mvc-without-endpoints</artifactId>
     <name>Integration tests for a Spring MVC project without using endpoints</name>

--- a/vaadin-spring-tests/test-spring-boot-contextpath/pom.xml
+++ b/vaadin-spring-tests/test-spring-boot-contextpath/pom.xml
@@ -5,7 +5,7 @@
     <parent>
         <groupId>com.vaadin</groupId>
         <artifactId>vaadin-spring-tests</artifactId>
-        <version>15.0-SNAPSHOT</version>
+        <version>16.0-SNAPSHOT</version>
     </parent>
     <artifactId>vaadin-test-spring-boot-contextpath</artifactId>
     <name>Vaadin Spring Boot integration tests when deployed using a context path</name>

--- a/vaadin-spring-tests/test-spring-boot-only-prepare/pom.xml
+++ b/vaadin-spring-tests/test-spring-boot-only-prepare/pom.xml
@@ -5,7 +5,7 @@
     <parent>
         <groupId>com.vaadin</groupId>
         <artifactId>vaadin-spring-tests</artifactId>
-        <version>15.0-SNAPSHOT</version>
+        <version>16.0-SNAPSHOT</version>
     </parent>
     <artifactId>vaadin-test-spring-boot-prepare</artifactId>
     <name>Vaadin Spring Boot integration tests with only prepare goal</name>

--- a/vaadin-spring-tests/test-spring-boot-scan/pom.xml
+++ b/vaadin-spring-tests/test-spring-boot-scan/pom.xml
@@ -6,7 +6,7 @@
     <parent>
         <groupId>com.vaadin</groupId>
         <artifactId>vaadin-spring-tests</artifactId>
-        <version>15.0-SNAPSHOT</version>
+        <version>16.0-SNAPSHOT</version>
     </parent>
     <artifactId>vaadin-test-spring-boot-scan</artifactId>
     <name>Vaadin Spring Boot integration tests with custom packages to scan</name>

--- a/vaadin-spring-tests/test-spring-boot/pom.xml
+++ b/vaadin-spring-tests/test-spring-boot/pom.xml
@@ -5,7 +5,7 @@
     <parent>
         <groupId>com.vaadin</groupId>
         <artifactId>vaadin-spring-tests</artifactId>
-        <version>15.0-SNAPSHOT</version>
+        <version>16.0-SNAPSHOT</version>
     </parent>
     <artifactId>vaadin-test-spring-boot</artifactId>
     <name>Vaadin Spring Boot integration tests</name>

--- a/vaadin-spring-tests/test-spring-common/pom.xml
+++ b/vaadin-spring-tests/test-spring-common/pom.xml
@@ -6,7 +6,7 @@
     <parent>
         <groupId>com.vaadin</groupId>
         <artifactId>vaadin-spring-tests</artifactId>
-        <version>15.0-SNAPSHOT</version>
+        <version>16.0-SNAPSHOT</version>
     </parent>
     <artifactId>vaadin-test-spring-common</artifactId>
     <name>Flow Spring Common UI components</name>

--- a/vaadin-spring-tests/test-spring-war/pom.xml
+++ b/vaadin-spring-tests/test-spring-war/pom.xml
@@ -5,7 +5,7 @@
     <parent>
         <groupId>com.vaadin</groupId>
         <artifactId>vaadin-spring-tests</artifactId>
-        <version>15.0-SNAPSHOT</version>
+        <version>16.0-SNAPSHOT</version>
     </parent>
     <artifactId>vaadin-test-spring-war</artifactId>
     <name>Vaadin Spring Boot deployable integration tests</name>

--- a/vaadin-spring-tests/test-spring-white-list/pom.xml
+++ b/vaadin-spring-tests/test-spring-white-list/pom.xml
@@ -22,7 +22,7 @@
     <parent>
         <groupId>com.vaadin</groupId>
         <artifactId>vaadin-spring-tests</artifactId>
-        <version>15.0-SNAPSHOT</version>
+        <version>16.0-SNAPSHOT</version>
     </parent>
     <artifactId>vaadin-test-spring-white-list</artifactId>
 

--- a/vaadin-spring-tests/test-ts-services-custom-client/pom.xml
+++ b/vaadin-spring-tests/test-ts-services-custom-client/pom.xml
@@ -5,7 +5,7 @@
     <parent>
         <groupId>com.vaadin</groupId>
         <artifactId>vaadin-spring-tests</artifactId>
-        <version>15.0-SNAPSHOT</version>
+        <version>16.0-SNAPSHOT</version>
     </parent>
     <artifactId>test-ts-services-custom-client</artifactId>
     <name>Integration tests for V15+ endpoints (custom Connect client)</name>

--- a/vaadin-spring-tests/test-ts-services/pom.xml
+++ b/vaadin-spring-tests/test-ts-services/pom.xml
@@ -5,7 +5,7 @@
     <parent>
         <groupId>com.vaadin</groupId>
         <artifactId>vaadin-spring-tests</artifactId>
-        <version>15.0-SNAPSHOT</version>
+        <version>16.0-SNAPSHOT</version>
     </parent>
     <artifactId>test-ts-services</artifactId>
     <name>Integration tests for V15+ endpoints</name>

--- a/vaadin-spring/pom.xml
+++ b/vaadin-spring/pom.xml
@@ -7,7 +7,7 @@
     <parent>
         <groupId>com.vaadin</groupId>
         <artifactId>vaadin-spring-parent</artifactId>
-        <version>15.0-SNAPSHOT</version>
+        <version>16.0-SNAPSHOT</version>
     </parent>
 
     <artifactId>vaadin-spring</artifactId>

--- a/vaadin-spring/src/main/java/com/vaadin/flow/spring/VaadinServletContextInitializer.java
+++ b/vaadin-spring/src/main/java/com/vaadin/flow/spring/VaadinServletContextInitializer.java
@@ -42,11 +42,6 @@ import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 import com.googlecode.gentyref.GenericTypeReflector;
-
-import com.vaadin.flow.internal.ReflectTools;
-import com.vaadin.flow.router.NotFoundException;
-import com.vaadin.flow.router.RouteNotFoundError;
-import com.vaadin.flow.server.startup.ServletDeployer;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.AnnotatedBeanDefinition;
@@ -67,10 +62,13 @@ import org.springframework.core.type.filter.AssignableTypeFilter;
 import com.vaadin.flow.component.Component;
 import com.vaadin.flow.component.WebComponentExporter;
 import com.vaadin.flow.function.DeploymentConfiguration;
+import com.vaadin.flow.internal.ReflectTools;
 import com.vaadin.flow.router.HasErrorParameter;
+import com.vaadin.flow.router.NotFoundException;
 import com.vaadin.flow.router.Route;
 import com.vaadin.flow.router.RouteAlias;
 import com.vaadin.flow.router.RouteConfiguration;
+import com.vaadin.flow.router.RouteNotFoundError;
 import com.vaadin.flow.server.AmbiguousRouteConfigurationException;
 import com.vaadin.flow.server.DeploymentConfigurationFactory;
 import com.vaadin.flow.server.DevModeHandler;
@@ -84,6 +82,8 @@ import com.vaadin.flow.server.startup.AnnotationValidator;
 import com.vaadin.flow.server.startup.ApplicationRouteRegistry;
 import com.vaadin.flow.server.startup.ClassLoaderAwareServletContainerInitializer;
 import com.vaadin.flow.server.startup.DevModeInitializer;
+import com.vaadin.flow.server.startup.LookupInitializer;
+import com.vaadin.flow.server.startup.ServletDeployer;
 import com.vaadin.flow.server.startup.ServletVerifier;
 import com.vaadin.flow.server.startup.VaadinAppShellInitializer;
 import com.vaadin.flow.server.startup.WebComponentConfigurationRegistryInitializer;
@@ -122,7 +122,7 @@ public class VaadinServletContextInitializer
             "org/aspectj", "org/bouncycastle", "org/dom4j", "org/easymock",
             "org/eclipse/persistence", "org/hamcrest", "org/hibernate",
             "org/javassist", "org/jboss", "org/jsoup", "org/seleniumhq",
-            "org/slf4j", "org/atmosphere", "org/springframework", 
+            "org/slf4j", "org/atmosphere", "org/springframework",
             "org/webjars/bowergithub", "org/yaml",
 
             "java/", "javax/", "javafx/", "com/sun/", "oracle/deploy",
@@ -132,9 +132,9 @@ public class VaadinServletContextInitializer
 
             "com/intellij/", "org/jetbrains").collect(Collectors.toList());
 
-   /**
-     * Packages that should be scanned by default and can't be overriden by
-     * a custom list.
+    /**
+     * Packages that should be scanned by default and can't be overriden by a
+     * custom list.
      */
     private static final List<String> DEFAULT_SCAN_ONLY = Stream
             .of(Component.class.getPackage().getName(),
@@ -200,6 +200,47 @@ public class VaadinServletContextInitializer
 
         void failFastContextInitialized(ServletContextEvent event)
                 throws ServletException;
+
+    }
+
+    private static class CompositeServletContextListener
+            implements ServletContextListener, Serializable {
+        private final List<FailFastServletContextListener> listeners = new ArrayList<>();
+
+        @Override
+        public void contextInitialized(ServletContextEvent event) {
+            listeners.forEach(listener -> listener.contextInitialized(event));
+        }
+
+        @Override
+        public void contextDestroyed(ServletContextEvent event) {
+            listeners.forEach(listener -> listener.contextDestroyed(event));
+        }
+
+        private void addListener(FailFastServletContextListener listener) {
+            listeners.add(listener);
+        }
+
+    }
+
+    private class LookupInitializerListener extends LookupInitializer
+            implements FailFastServletContextListener {
+
+        @Override
+        public void failFastContextInitialized(ServletContextEvent event)
+                throws ServletException {
+            // intentionally make annotation unmodifiable empty list because
+            // LookupInitializer doesn't have annotations at the moment
+            List<Class<? extends Annotation>> annotations = Collections
+                    .emptyList();
+            List<Class<?>> types = new ArrayList<>();
+            collectHandleTypes(LookupInitializer.class, annotations, types);
+            Set<Class<?>> classes = findByAnnotationOrSuperType(
+                    getDefaultPackages(), appContext, Collections.emptyList(),
+                    types).collect(Collectors.toSet());
+            process(classes, event.getServletContext());
+        }
+
     }
 
     private class RouteServletContextListener
@@ -299,12 +340,14 @@ public class VaadinServletContextInitializer
             Set<Class<? extends Component>> errorComponents = findBySuperType(
                     getErrorParameterPackages(), HasErrorParameter.class)
                             .filter(Component.class::isAssignableFrom)
-                            // Replace Flow default with custom version for Spring
+                            // Replace Flow default with custom version for
+                            // Spring
                             .filter(clazz -> clazz != RouteNotFoundError.class)
                             .map(clazz -> (Class<? extends Component>) clazz)
-                    .collect(Collectors.toSet());
+                            .collect(Collectors.toSet());
 
-            // If there is no custom HasErrorParameter<? super NotFoundException>
+            // If there is no custom HasErrorParameter<? super
+            // NotFoundException>
             // add SpringRouteNotFoundError, with Spring Boot specific hints
             if (errorComponents.stream().noneMatch(clazz -> {
                 Class<?> exceptionType = ReflectTools.getGenericInterfaceType(
@@ -362,9 +405,9 @@ public class VaadinServletContextInitializer
                     .createDeploymentConfiguration(this.getClass(), event,
                             appContext);
 
-            if (config == null || config.isProductionMode() ||
-                    !config.enableDevServer() ||
-                    isDevModeAlreadyStarted(event.getServletContext())) {
+            if (config == null || config.isProductionMode()
+                    || !config.enableDevServer()
+                    || isDevModeAlreadyStarted(event.getServletContext())) {
                 return;
             }
 
@@ -379,7 +422,8 @@ public class VaadinServletContextInitializer
 
             List<Class<? extends Annotation>> annotations = new ArrayList<>();
             List<Class<?>> superTypes = new ArrayList<>();
-            collectDevModeTypes(annotations, superTypes);
+            collectHandleTypes(DevModeInitializer.class, annotations,
+                    superTypes);
 
             Set<Class<?>> classes = findByAnnotationOrSuperType(basePackages,
                     customLoader, annotations, superTypes)
@@ -420,25 +464,8 @@ public class VaadinServletContextInitializer
             return npmPackages;
         }
 
-        @SuppressWarnings("unchecked")
-        private void collectDevModeTypes(
-                List<Class<? extends Annotation>> annotations,
-                List<Class<?>> superTypes) {
-            HandlesTypes handleTypes = DevModeInitializer.class
-                    .getAnnotation(HandlesTypes.class);
-            assert handleTypes != null;
-            for (Class<?> clazz : handleTypes.value()) {
-                if (clazz.isAnnotation()) {
-                    annotations.add((Class<? extends Annotation>) clazz);
-                } else {
-                    superTypes.add(clazz);
-                }
-            }
-        }
-
         private boolean isScanOnlySet() {
-            return customScanOnly != null
-                    && !customScanOnly.isEmpty();
+            return customScanOnly != null && !customScanOnly.isEmpty();
         }
 
         private boolean isDevModeAlreadyStarted(ServletContext servletContext) {
@@ -551,10 +578,21 @@ public class VaadinServletContextInitializer
         // Verify servlet version also for SpringBoot.
         ServletVerifier.verifyServletVersion();
 
-        servletContext.addListener(new VaadinAppShellContextListener());
+        VaadinServletContext vaadinContext = new VaadinServletContext(
+                servletContext);
+        servletContext.addListener(createCompositeListener(vaadinContext));
+    }
+
+    private CompositeServletContextListener createCompositeListener(
+            VaadinServletContext context) {
+        CompositeServletContextListener compositeListener = new CompositeServletContextListener();
+
+        compositeListener.addListener(new LookupInitializerListener());
+
+        compositeListener.addListener(new VaadinAppShellContextListener());
 
         ApplicationRouteRegistry registry = ApplicationRouteRegistry
-                .getInstance(new VaadinServletContext(servletContext));
+                .getInstance(context);
 
         // If the registry is already initialized then RouteRegistryInitializer
         // has done its job already, skip the custom routes search
@@ -568,24 +606,26 @@ public class VaadinServletContextInitializer
              * because an RouteRegistryInitializer has not been executed (end
              * never will).
              */
-            servletContext.addListener(new RouteServletContextListener());
+            compositeListener.addListener(new RouteServletContextListener());
         }
 
-        servletContext.addListener(new ErrorParameterServletContextListener());
+        compositeListener
+                .addListener(new ErrorParameterServletContextListener());
 
-        servletContext
+        compositeListener
                 .addListener(new AnnotationValidatorServletContextListener());
 
-        servletContext.addListener(new DevModeServletContextListener());
+        compositeListener.addListener(new DevModeServletContextListener());
 
         // Skip custom web component builders search if registry already
         // initialized
-        if (!WebComponentConfigurationRegistry
-                .getInstance(new VaadinServletContext(servletContext))
+        if (!WebComponentConfigurationRegistry.getInstance(context)
                 .hasConfigurations()) {
-            servletContext
+            compositeListener
                     .addListener(new WebComponentServletContextListener());
         }
+
+        return compositeListener;
     }
 
     private Stream<Class<?>> findByAnnotation(Collection<String> packages,
@@ -649,9 +689,11 @@ public class VaadinServletContextInitializer
     }
 
     private Collection<String> getErrorParameterPackages() {
-        return Stream.concat(
-                Stream.of(HasErrorParameter.class.getPackage().getName()),
-                getDefaultPackages().stream()).collect(Collectors.toSet());
+        return Stream
+                .concat(Stream
+                        .of(HasErrorParameter.class.getPackage().getName()),
+                        getDefaultPackages().stream())
+                .collect(Collectors.toSet());
     }
 
     private List<String> getDefaultPackages() {
@@ -673,6 +715,20 @@ public class VaadinServletContextInitializer
         return packagesList;
     }
 
+    private static void collectHandleTypes(Class<?> clazz,
+            List<Class<? extends Annotation>> annotations,
+            List<Class<?>> superTypes) {
+        HandlesTypes handleTypes = clazz.getAnnotation(HandlesTypes.class);
+        assert handleTypes != null;
+        for (Class<?> type : handleTypes.value()) {
+            if (type.isAnnotation()) {
+                annotations.add((Class<? extends Annotation>) type);
+            } else {
+                superTypes.add(type);
+            }
+        }
+    }
+
     /**
      * For NPM we scan all packages. For performance reasons and due to problems
      * with atmosphere we skip known packaged from our resources collection.
@@ -682,10 +738,9 @@ public class VaadinServletContextInitializer
 
         private final PrefixTree scanNever = new PrefixTree(DEFAULT_SCAN_NEVER);
 
-        private final PrefixTree scanAlways = new PrefixTree(
-                DEFAULT_SCAN_ONLY.stream()
-                        .map(packageName -> packageName.replace('.', '/'))
-                        .collect(Collectors.toList()));
+        private final PrefixTree scanAlways = new PrefixTree(DEFAULT_SCAN_ONLY
+                .stream().map(packageName -> packageName.replace('.', '/'))
+                .collect(Collectors.toList()));
 
         public CustomResourceLoader(ResourceLoader resourceLoader,
                 List<String> addedScanNever) {
@@ -762,8 +817,7 @@ public class VaadinServletContextInitializer
         }
 
         private boolean shouldPathBeScanned(String path) {
-            return scanAlways.hasPrefix(path)
-                    || !scanNever.hasPrefix(path);
+            return scanAlways.hasPrefix(path) || !scanNever.hasPrefix(path);
         }
     }
 

--- a/vaadin-spring/src/main/java/com/vaadin/flow/spring/VaadinServletContextInitializer.java
+++ b/vaadin-spring/src/main/java/com/vaadin/flow/spring/VaadinServletContextInitializer.java
@@ -61,6 +61,7 @@ import org.springframework.core.type.filter.AssignableTypeFilter;
 
 import com.vaadin.flow.component.Component;
 import com.vaadin.flow.component.WebComponentExporter;
+import com.vaadin.flow.di.Lookup;
 import com.vaadin.flow.function.DeploymentConfiguration;
 import com.vaadin.flow.internal.ReflectTools;
 import com.vaadin.flow.router.HasErrorParameter;
@@ -229,6 +230,12 @@ public class VaadinServletContextInitializer
         @Override
         public void failFastContextInitialized(ServletContextEvent event)
                 throws ServletException {
+            VaadinServletContext vaadinContext = new VaadinServletContext(
+                    event.getServletContext());
+            if (vaadinContext.getAttribute(Lookup.class) != null) {
+                return;
+            }
+
             // intentionally make annotation unmodifiable empty list because
             // LookupInitializer doesn't have annotations at the moment
             List<Class<? extends Annotation>> annotations = Collections

--- a/vaadin-spring/src/test/java/com/vaadin/flow/spring/SpringStubServletConfigTest.java
+++ b/vaadin-spring/src/test/java/com/vaadin/flow/spring/SpringStubServletConfigTest.java
@@ -8,11 +8,14 @@ import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
 import org.mockito.Mock;
+import org.mockito.Mockito;
 import org.mockito.MockitoAnnotations;
 import org.springframework.boot.web.servlet.ServletRegistrationBean;
 import org.springframework.context.ApplicationContext;
 import org.springframework.core.env.Environment;
 
+import com.vaadin.flow.di.Lookup;
+import com.vaadin.flow.di.ResourceProvider;
 import com.vaadin.flow.function.DeploymentConfiguration;
 import com.vaadin.flow.server.Constants;
 
@@ -38,6 +41,14 @@ public class SpringStubServletConfigTest {
     @Before
     public void init() {
         MockitoAnnotations.initMocks(this);
+
+        Lookup lookup = Mockito.mock(Lookup.class);
+        ResourceProvider provider = Mockito.mock(ResourceProvider.class);
+        Mockito.when(lookup.lookup(ResourceProvider.class))
+                .thenReturn(provider);
+
+        Mockito.when(context.getAttribute(Lookup.class.getName()))
+                .thenReturn(lookup);
 
         when(applicationContext.getBean(Environment.class))
                 .thenReturn(environment);

--- a/vaadin-spring/src/test/java/com/vaadin/flow/spring/VaadinServletContextInitializerTest.java
+++ b/vaadin-spring/src/test/java/com/vaadin/flow/spring/VaadinServletContextInitializerTest.java
@@ -1,20 +1,15 @@
 package com.vaadin.flow.spring;
 
+import javax.servlet.ServletContext;
+import javax.servlet.ServletContextEvent;
+import javax.servlet.ServletContextListener;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Stream;
+
 import com.google.common.collect.Maps;
-
-import com.vaadin.flow.component.Component;
-import com.vaadin.flow.function.DeploymentConfiguration;
-import com.vaadin.flow.router.BeforeEnterEvent;
-import com.vaadin.flow.router.ErrorParameter;
-import com.vaadin.flow.router.HasErrorParameter;
-import com.vaadin.flow.router.NotFoundException;
-import com.vaadin.flow.router.RouteNotFoundError;
-import com.vaadin.flow.server.VaadinServletContext;
-import com.vaadin.flow.server.startup.ApplicationRouteRegistry;
-import com.vaadin.flow.server.startup.DevModeInitializer;
-import com.vaadin.flow.server.startup.ServletDeployer;
-import com.vaadin.flow.spring.router.SpringRouteNotFoundError;
-
 import org.apache.commons.lang3.StringUtils;
 import org.junit.Assert;
 import org.junit.Before;
@@ -29,21 +24,25 @@ import org.springframework.boot.autoconfigure.AutoConfigurationPackages;
 import org.springframework.context.ApplicationContext;
 import org.springframework.core.env.Environment;
 
-import javax.servlet.ServletContext;
-import javax.servlet.ServletContextEvent;
-import javax.servlet.ServletContextListener;
-import java.util.Collections;
-import java.util.Map;
-import java.util.concurrent.atomic.AtomicReference;
-import java.util.stream.Stream;
+import com.vaadin.flow.component.Component;
+import com.vaadin.flow.function.DeploymentConfiguration;
+import com.vaadin.flow.router.BeforeEnterEvent;
+import com.vaadin.flow.router.ErrorParameter;
+import com.vaadin.flow.router.HasErrorParameter;
+import com.vaadin.flow.router.NotFoundException;
+import com.vaadin.flow.router.RouteNotFoundError;
+import com.vaadin.flow.server.VaadinServletContext;
+import com.vaadin.flow.server.startup.ApplicationRouteRegistry;
+import com.vaadin.flow.server.startup.DevModeInitializer;
+import com.vaadin.flow.server.startup.ServletDeployer;
+import com.vaadin.flow.spring.router.SpringRouteNotFoundError;
 
 @RunWith(PowerMockRunner.class)
-@PrepareForTest({DevModeInitializer.class,
+@PrepareForTest({ DevModeInitializer.class,
         VaadinServletContextInitializer.SpringStubServletConfig.class,
-        VaadinServletContextInitializer.class,
-        ServletDeployer.class,
+        VaadinServletContextInitializer.class, ServletDeployer.class,
         ServletDeployer.StubServletConfig.class,
-        AutoConfigurationPackages.class})
+        AutoConfigurationPackages.class })
 public class VaadinServletContextInitializerTest {
 
     @Mock
@@ -60,7 +59,8 @@ public class VaadinServletContextInitializerTest {
 
     @Before
     public void init() {
-        PowerMockito.mockStatic(VaadinServletContextInitializer.SpringStubServletConfig.class);
+        PowerMockito.mockStatic(
+                VaadinServletContextInitializer.SpringStubServletConfig.class);
         PowerMockito.mockStatic(ServletDeployer.class);
         PowerMockito.mockStatic(ServletDeployer.StubServletConfig.class);
         PowerMockito.mockStatic(DevModeInitializer.class);
@@ -68,30 +68,34 @@ public class VaadinServletContextInitializerTest {
     }
 
     @Test
-    public void onStartup_devModeNotInitialized_devModeInitialized() throws Exception {
+    public void onStartup_devModeNotInitialized_devModeInitialized()
+            throws Exception {
         initDefaultMocks();
 
-        VaadinServletContextInitializer vaadinServletContextInitializer =
-                getStubbedVaadinServletContextInitializer();
+        VaadinServletContextInitializer vaadinServletContextInitializer = getStubbedVaadinServletContextInitializer();
 
         // Simulate Spring context start only
         vaadinServletContextInitializer.onStartup(servletContext);
 
-        // This is how PowerMockito works, call PowerMockito.verifyStatic() first
+        // This is how PowerMockito works, call PowerMockito.verifyStatic()
+        // first
         // to start verifying behavior of DevModeInitializer static methods
         PowerMockito.verifyStatic(DevModeInitializer.class);
-        // IMPORTANT:  Call the static method we want to verify.
-        // In our case, we want to check if Dev Mode has been started within onStartup() call,
-        // that means DevModeInitializer.initDevModeHandler() should has been called exactly one time
-        DevModeInitializer.initDevModeHandler(Mockito.any(), Mockito.any(), Mockito.any());
+        // IMPORTANT: Call the static method we want to verify.
+        // In our case, we want to check if Dev Mode has been started within
+        // onStartup() call,
+        // that means DevModeInitializer.initDevModeHandler() should has been
+        // called exactly one time
+        DevModeInitializer.initDevModeHandler(Mockito.any(), Mockito.any(),
+                Mockito.any());
     }
 
     @Test
-    public void onStartup_devModeAlreadyInitialized_devModeInitializationSkipped() throws Exception {
+    public void onStartup_devModeAlreadyInitialized_devModeInitializationSkipped()
+            throws Exception {
         initDefaultMocks();
 
-        VaadinServletContextInitializer vaadinServletContextInitializer =
-                getStubbedVaadinServletContextInitializer();
+        VaadinServletContextInitializer vaadinServletContextInitializer = getStubbedVaadinServletContextInitializer();
 
         DevModeInitializer devModeInitializer = getStubbedDevModeInitializer();
 
@@ -99,22 +103,27 @@ public class VaadinServletContextInitializerTest {
         devModeInitializer.process(Collections.emptySet(), servletContext);
         vaadinServletContextInitializer.onStartup(servletContext);
 
-        // This is how PowerMockito works, call PowerMockito.verifyStatic() first
+        // This is how PowerMockito works, call PowerMockito.verifyStatic()
+        // first
         // to start verifying behavior of DevModeInitializer static methods
         PowerMockito.verifyStatic(DevModeInitializer.class);
-        // IMPORTANT:  Call the static method we want to verify.
+        // IMPORTANT: Call the static method we want to verify.
         // In our case, we want to check if Dev Mode has been started within
-        // devModeInitializer.process() call (i.e. from Servlet Container), and not started again
+        // devModeInitializer.process() call (i.e. from Servlet Container), and
+        // not started again
         // within DevModeInitializer.initDevModeHandler() (Spring context),
         // so, we expect this method has been called exactly one time:
-        DevModeInitializer.initDevModeHandler(Mockito.any(), Mockito.any(), Mockito.any());
+        DevModeInitializer.initDevModeHandler(Mockito.any(), Mockito.any(),
+                Mockito.any());
     }
 
     @Test
-    public void errorParameterServletContextListenerEvent_defaultRouteNotFoundView_defaultRouteNotFoundViewIsRegistered() throws Exception {
+    public void errorParameterServletContextListenerEvent_defaultRouteNotFoundView_defaultRouteNotFoundViewIsRegistered()
+            throws Exception {
         // given
         initDefaultMocks();
-        Runnable when = initRouteNotFoundMocksAndGetContextInitializedMockCall(getStubbedVaadinServletContextInitializer());
+        Runnable when = initRouteNotFoundMocksAndGetContextInitializedMockCall(
+                getStubbedVaadinServletContextInitializer());
 
         // when
         when.run();
@@ -122,28 +131,27 @@ public class VaadinServletContextInitializerTest {
         // then
         ApplicationRouteRegistry registry = ApplicationRouteRegistry
                 .getInstance(new VaadinServletContext(servletContext));
-        final Class<? extends Component> navigationTarget =
-                registry.getErrorNavigationTarget(new NotFoundException()).get().getNavigationTarget();
+        final Class<? extends Component> navigationTarget = registry
+                .getErrorNavigationTarget(new NotFoundException()).get()
+                .getNavigationTarget();
         Assert.assertEquals(SpringRouteNotFoundError.class, navigationTarget);
     }
 
     @Test
-    public void errorParameterServletContextListenerEvent_hasCustomRouteNotFoundViewExtendingRouteNotFoundError_customRouteNotFoundViewIsRegistered() throws Exception {
+    public void errorParameterServletContextListenerEvent_hasCustomRouteNotFoundViewExtendingRouteNotFoundError_customRouteNotFoundViewIsRegistered()
+            throws Exception {
         // given
         initDefaultMocks();
         VaadinServletContextInitializer initializer = getStubbedVaadinServletContextInitializer();
-        Runnable when = initRouteNotFoundMocksAndGetContextInitializedMockCall(initializer);
+        Runnable when = initRouteNotFoundMocksAndGetContextInitializedMockCall(
+                initializer);
 
         class TestErrorView extends RouteNotFoundError {
         }
 
-        PowerMockito.doReturn(Stream.of(TestErrorView.class))
-                .when(initializer,
-                        "findByAnnotationOrSuperType",
-                        Mockito.anyCollection(),
-                        Mockito.any(),
-                        Mockito.anyCollection(),
-                        Mockito.anyCollection());
+        PowerMockito.doAnswer(invocation -> Stream.of(TestErrorView.class))
+                .when(initializer, "findBySuperType", Mockito.anyCollection(),
+                        Mockito.eq(HasErrorParameter.class));
 
         // when
         when.run();
@@ -151,32 +159,33 @@ public class VaadinServletContextInitializerTest {
         // then
         ApplicationRouteRegistry registry = ApplicationRouteRegistry
                 .getInstance(new VaadinServletContext(servletContext));
-        final Class<? extends Component> navigationTarget =
-                registry.getErrorNavigationTarget(new NotFoundException()).get().getNavigationTarget();
+        final Class<? extends Component> navigationTarget = registry
+                .getErrorNavigationTarget(new NotFoundException()).get()
+                .getNavigationTarget();
         Assert.assertEquals(TestErrorView.class, navigationTarget);
     }
 
     @Test
-    public void errorParameterServletContextListenerEvent_hasCustomRouteNotFoundViewImplementingHasErrorParameter_customRouteNotFoundViewIsRegistered() throws Exception {
+    public void errorParameterServletContextListenerEvent_hasCustomRouteNotFoundViewImplementingHasErrorParameter_customRouteNotFoundViewIsRegistered()
+            throws Exception {
         // given
         initDefaultMocks();
         VaadinServletContextInitializer initializer = getStubbedVaadinServletContextInitializer();
-        Runnable when = initRouteNotFoundMocksAndGetContextInitializedMockCall(initializer);
+        Runnable when = initRouteNotFoundMocksAndGetContextInitializedMockCall(
+                initializer);
 
-        class TestErrorView extends Component implements HasErrorParameter<NotFoundException> {
+        class TestErrorView extends Component
+                implements HasErrorParameter<NotFoundException> {
             @Override
-            public int setErrorParameter(BeforeEnterEvent event, ErrorParameter<NotFoundException> parameter) {
+            public int setErrorParameter(BeforeEnterEvent event,
+                    ErrorParameter<NotFoundException> parameter) {
                 return 0;
             }
         }
 
-        PowerMockito.doReturn(Stream.of(TestErrorView.class))
-                .when(initializer,
-                        "findByAnnotationOrSuperType",
-                        Mockito.anyCollection(),
-                        Mockito.any(),
-                        Mockito.anyCollection(),
-                        Mockito.anyCollection());
+        PowerMockito.doAnswer(invocation -> Stream.of(TestErrorView.class))
+                .when(initializer, "findBySuperType", Mockito.anyCollection(),
+                        Mockito.eq(HasErrorParameter.class));
 
         // when
         when.run();
@@ -184,8 +193,9 @@ public class VaadinServletContextInitializerTest {
         // then
         ApplicationRouteRegistry registry = ApplicationRouteRegistry
                 .getInstance(new VaadinServletContext(servletContext));
-        final Class<? extends Component> navigationTarget =
-                registry.getErrorNavigationTarget(new NotFoundException()).get().getNavigationTarget();
+        final Class<? extends Component> navigationTarget = registry
+                .getErrorNavigationTarget(new NotFoundException()).get()
+                .getNavigationTarget();
         Assert.assertEquals(TestErrorView.class, navigationTarget);
     }
 
@@ -197,22 +207,23 @@ public class VaadinServletContextInitializerTest {
         AtomicReference<ServletContextListener> theListener = new AtomicReference<>();
         Mockito.doAnswer(answer -> {
             ServletContextListener listener = answer.getArgument(0);
-            if ("ErrorParameterServletContextListener".equals(
-                    listener.getClass().getSimpleName())) {
-                theListener.set(listener);
-            }
+            theListener.set(listener);
             return null;
-        }).when(servletContext).addListener(Mockito.any(ServletContextListener.class));
+        }).when(servletContext)
+                .addListener(Mockito.any(ServletContextListener.class));
 
         vaadinServletContextInitializer.onStartup(servletContext);
 
-        Mockito.when(applicationContext.getBeanNamesForType(VaadinScanPackagesRegistrar.VaadinScanPackages.class))
-                .thenReturn(new String[]{});
+        Mockito.when(applicationContext.getBeanNamesForType(
+                VaadinScanPackagesRegistrar.VaadinScanPackages.class))
+                .thenReturn(new String[] {});
         PowerMockito.when(AutoConfigurationPackages.class, "has",
                 applicationContext).thenReturn(false);
 
-        ServletContextEvent initEventMock = Mockito.mock(ServletContextEvent.class);
-        Mockito.when(initEventMock.getServletContext()).thenReturn(servletContext);
+        ServletContextEvent initEventMock = Mockito
+                .mock(ServletContextEvent.class);
+        Mockito.when(initEventMock.getServletContext())
+                .thenReturn(servletContext);
 
         return () -> {
             theListener.get().contextInitialized(initEventMock);
@@ -221,51 +232,45 @@ public class VaadinServletContextInitializerTest {
 
     private DevModeInitializer getStubbedDevModeInitializer() throws Exception {
         PowerMockito.when(ServletDeployer.StubServletConfig.class,
-                "createDeploymentConfiguration",
-                Mockito.any(),
-                Mockito.any())
+                "createDeploymentConfiguration", Mockito.any(), Mockito.any())
                 .thenReturn(deploymentConfiguration);
 
-        PowerMockito.when(DevModeInitializer.class,
-                "isDevModeAlreadyStarted",
-                servletContext)
-                .thenCallRealMethod();
+        PowerMockito.when(DevModeInitializer.class, "isDevModeAlreadyStarted",
+                servletContext).thenCallRealMethod();
 
         return new DevModeInitializer();
     }
 
-    private VaadinServletContextInitializer getStubbedVaadinServletContextInitializer() throws Exception {
-        VaadinServletContextInitializer vaadinServletContextInitializerMock =
-                PowerMockito.spy(new VaadinServletContextInitializer(applicationContext));
+    private VaadinServletContextInitializer getStubbedVaadinServletContextInitializer()
+            throws Exception {
+        VaadinServletContextInitializer vaadinServletContextInitializerMock = PowerMockito
+                .spy(new VaadinServletContextInitializer(applicationContext));
 
-        PowerMockito.when(VaadinServletContextInitializer.SpringStubServletConfig.class,
-                "createDeploymentConfiguration",
-                Mockito.any(),
-                Mockito.any(),
-                Mockito.any())
-                .thenReturn(deploymentConfiguration);
+        PowerMockito.when(
+                VaadinServletContextInitializer.SpringStubServletConfig.class,
+                "createDeploymentConfiguration", Mockito.any(), Mockito.any(),
+                Mockito.any()).thenReturn(deploymentConfiguration);
 
-        PowerMockito.doReturn(Stream.empty())
-                .when(vaadinServletContextInitializerMock,
-                        "findByAnnotationOrSuperType",
-                        Mockito.anyCollection(),
-                        Mockito.any(),
-                        Mockito.anyCollection(),
-                        Mockito.anyCollection());
+        PowerMockito.doAnswer(invocation -> Stream.empty()).when(
+                vaadinServletContextInitializerMock,
+                "findByAnnotationOrSuperType", Mockito.anyCollection(),
+                Mockito.any(), Mockito.anyCollection(),
+                Mockito.anyCollection());
 
-        Mockito.doAnswer(answer -> {
-            ServletContextListener devModeListener =
-                    answer.getArgument(0);
-            if ("DevModeServletContextListener".equals(
-                    devModeListener.getClass().getSimpleName())) {
-                devModeListener.contextInitialized(new ServletContextEvent(servletContext));
-            }
+        PowerMockito.doReturn(Collections.emptyList()).when(
+                vaadinServletContextInitializerMock, "getDefaultPackages");
+
+        Mockito.doAnswer(invocation -> {
+            ServletContextListener devModeListener = invocation.getArgument(0);
+            devModeListener.contextInitialized(
+                    new ServletContextEvent(servletContext));
             return null;
-        }).when(servletContext).addListener(Mockito.any(ServletContextListener.class));
+        }).when(servletContext)
+                .addListener(Mockito.any(ServletContextListener.class));
 
-        PowerMockito.doNothing()
-                .when(ServletDeployer.class);
-        ServletDeployer.logAppStartupToConsole(Mockito.any(), Mockito.anyBoolean());
+        PowerMockito.doNothing().when(ServletDeployer.class);
+        ServletDeployer.logAppStartupToConsole(Mockito.any(),
+                Mockito.anyBoolean());
 
         return vaadinServletContextInitializerMock;
     }
@@ -278,18 +283,18 @@ public class VaadinServletContextInitializerTest {
     }
 
     private void mockServletContext() {
-        final Map<String, Object> servletContextAttributesMap = Maps.newHashMap();
+        final Map<String, Object> servletContextAttributesMap = Maps
+                .newHashMap();
         Mockito.doAnswer(answer -> {
-                    String key = answer.getArgument(0, String.class);
-                    Object value = answer.getArgument(1, Object.class);
-                    servletContextAttributesMap.putIfAbsent(key, value);
-                    return null;
-                })
-                .when(servletContext)
-                .setAttribute(Mockito.anyString(), Mockito.any());
+            String key = answer.getArgument(0, String.class);
+            Object value = answer.getArgument(1, Object.class);
+            servletContextAttributesMap.putIfAbsent(key, value);
+            return null;
+        }).when(servletContext).setAttribute(Mockito.anyString(),
+                Mockito.any());
         Mockito.when(servletContext.getAttribute(Mockito.anyString()))
-                .thenAnswer(answer ->
-                        servletContextAttributesMap.get(answer.getArgument(0, String.class)));
+                .thenAnswer(answer -> servletContextAttributesMap
+                        .get(answer.getArgument(0, String.class)));
         Mockito.when(servletContext.getServletRegistrations())
                 .thenReturn(Maps.newHashMap());
     }
@@ -307,6 +312,7 @@ public class VaadinServletContextInitializerTest {
     }
 
     private void mockApplicationContext() {
-        Mockito.when(applicationContext.getEnvironment()).thenReturn(environment);
+        Mockito.when(applicationContext.getEnvironment())
+                .thenReturn(environment);
     }
 }

--- a/vaadin-spring/src/test/java/com/vaadin/flow/spring/instantiator/SpringInstantiatorTest.java
+++ b/vaadin-spring/src/test/java/com/vaadin/flow/spring/instantiator/SpringInstantiatorTest.java
@@ -15,16 +15,16 @@
  */
 package com.vaadin.flow.spring.instantiator;
 
+import javax.servlet.ServletConfig;
+import javax.servlet.ServletContext;
+import javax.servlet.ServletException;
+
 import java.util.Collections;
 import java.util.List;
 import java.util.Locale;
 import java.util.Properties;
 import java.util.Set;
 import java.util.stream.Collectors;
-
-import javax.servlet.ServletConfig;
-import javax.servlet.ServletContext;
-import javax.servlet.ServletException;
 
 import org.junit.Assert;
 import org.junit.Test;
@@ -43,9 +43,10 @@ import org.springframework.test.context.junit4.SpringRunner;
 
 import com.vaadin.flow.component.html.Div;
 import com.vaadin.flow.di.Instantiator;
+import com.vaadin.flow.di.Lookup;
+import com.vaadin.flow.di.ResourceProvider;
 import com.vaadin.flow.function.DeploymentConfiguration;
 import com.vaadin.flow.i18n.I18NProvider;
-import com.vaadin.flow.server.Constants;
 import com.vaadin.flow.server.ServiceInitEvent;
 import com.vaadin.flow.server.VaadinServiceInitListener;
 import com.vaadin.flow.server.VaadinServletService;
@@ -158,12 +159,13 @@ public class SpringInstantiatorTest {
     }
 
     public static VaadinServletService getService(ApplicationContext context,
-            Properties configProperties ) throws ServletException {
+            Properties configProperties) throws ServletException {
         return getService(context, configProperties, false);
     }
 
     public static VaadinServletService getService(ApplicationContext context,
-            Properties configProperties, boolean rootMapping) throws ServletException {
+            Properties configProperties, boolean rootMapping)
+            throws ServletException {
         SpringServlet servlet = new SpringServlet(context, rootMapping) {
             @Override
             protected DeploymentConfiguration createDeploymentConfiguration(
@@ -179,6 +181,13 @@ public class SpringInstantiatorTest {
 
         ServletConfig config = Mockito.mock(ServletConfig.class);
         ServletContext servletContext = Mockito.mock(ServletContext.class);
+
+        Lookup lookup = Mockito.mock(Lookup.class);
+        ResourceProvider provider = Mockito.mock(ResourceProvider.class);
+        Mockito.when(lookup.lookup(ResourceProvider.class))
+                .thenReturn(provider);
+        Mockito.when(servletContext.getAttribute(Lookup.class.getName()))
+                .thenReturn(lookup);
 
         Mockito.when(config.getServletContext()).thenReturn(servletContext);
 

